src/ol/Collection.js

```diff
@@ -278,14 +288,18 @@ class Collection extends BaseObject {
       const prev = this.array_[index];
       this.array_[index] = elem;
       this.dispatchEvent(
-        new CollectionEvent(CollectionEventType.REMOVE, prev, index)
+        /** @type {CollectionEvent<T>} */ (
+          new CollectionEvent(CollectionEventType.REMOVE, prev, index)
+        )
       );
       this.dispatchEvent(
-        new CollectionEvent(CollectionEventType.ADD, elem, index)
+        /** @type {CollectionEvent<T>} */ (
+          new CollectionEvent(CollectionEventType.ADD, elem, index)
+        )
       );
     } else {
-      for (let j = n; j < index; ++j) {
-        this.insertAt(j, undefined);
+      if (index > n) {
+        throw new Error('Index out of bounds: ' + index);
       }
```

Comment:
Maybe worth mentioning in the upgrade notes that you can no longer insert elements at index values beyond the length of the collection (leaving a sparse collection).